### PR TITLE
[#12775] feat(core): Add OCC for tag metadata

### LINKED ISSUE #12775: [Subtask] Implement version-CAS OCC for tag
### Describe the subtask

Implement database optimistic concurrency control and atomic service operations for managed tags. Split from #12347.

- Audit tag create, alter, delete, assignment-value, and metadata-object association mutations.
- Fence the active parent metalake during tag creation using the shared parent-lock pattern.
- Advance `current_version` on every successful alter.
- Match updates by tag ID, expected version, and active-row state.
- Preserve monotonic versions during overwrite instead of resetting the OCC token.
- Make tag deletion version checked and execute the root CAS before relationship, ownership, and change-log cleanup.
- Use `OccWriteSupport` to classify missing entities versus optimistic-lock conflicts.
- Apply association additions, removals, and value changes atomically; a conflict must not leave a partial assignment set.
- Add H2, MySQL, and PostgreSQL coverage for create/drop fencing, alter races, stale deletes, overwrite versioning, assignment updates, and rollback.

No database migration is required because the OCC columns already exist.

Depends on #12341 and #12639.

### Parent issue

https://github.com/apache/gravitino/issues/12166

### PR DESCRIPTION
**Pull request by yuqi1129**
### What changes were proposed in this pull request?

- Apply stable-ID and expected-version CAS to tag alter and delete operations.
- Replace database-specific tag overwrite upserts with locked, monotonic root updates that preserve the stored tag ID.
- Lock parent metalakes during creation and tag roots during assignment changes.
- Keep tag deletion and all dependent relationship cleanup, including policy-on-tag rows, atomic.
- Add real two-transaction interleaving coverage for create/delete, overwrite/rename, and assignment/rename races.

### Why are the changes needed?

Concurrent tag writes could otherwise cause lost updates, stale deletions, ID/version resets, or partial relationship cleanup.

Fix: #12775

### Does this PR introduce _any_ user-facing change?

No API or configuration change. Concurrent stale writes now fail deterministically; a tag deleted or renamed by a competing transaction preserves the existing idempotent delete result.

### How was this patch tested?

- `./gradlew :core:spotlessApply :core:compileTestJava :core:javadoc`
- `./gradlew :core:test --tests org.apache.gravitino.storage.relational.service.TestTagMetaService --tests org.apache.gravitino.storage.relational.service.TestPolicyTagRelService --tests org.apache.gravitino.tag.TestTagManager -PskipDockerTests=false`
- `TestTagMetaService` ran against H2, MySQL, and PostgreSQL.

### CODE REVIEW

File: core/src/main/java/org/apache/gravitino/storage/relational/service/TagMetaService.java [hunk 1]
```diff
@@ -156,27 +174,47 @@ public <E extends Entity & HasIdentifier> TagEntity updateTag(
...
+    } catch (NoSuchEntityException e) {
+      return false;
+    }
+    return deleteTag(identifier, tagPO);
+  }
+
+  boolean deleteTag(NameIdentifier identifier, TagPO tagPO) {
+    long tagId = tagPO.getTagId();
 
     SessionUtils.doMultipleWithCommit(
```
Comment (jerryshao) [inline core/src/main/java/org/apache/gravitino/storage/relational/service/TagMetaService.java]: **altitude**: `deleteTag`'s five cascade cleanup calls (tag-metadata-object rel, policy-tag rel, policy-metadata-object rel, owner rel, securable-object rel) are fired via plain `SessionUtils.doWithoutCommit` with no row-count verification, relying entirely on the undocumented invariant that `deleteTagWithVersion` runs first in this `doMultipleWithCommit` varargs list — unlike sibling services (e.g. `SchemaMetaService`), which route cascade cleanup through `OccWriteSupport.deleteChildrenWithVersions` to assert the affected row count.

Correctness of the whole cascade depends on `deleteTagWithVersion` (line 190) staying the first element of the varargs list, with nothing in the code or a comment flagging that requirement. If a future refactor reorders the suppliers, adds an early cascade step before the version-CAS, or copies this pattern into a context where cleanup runs in a separately-committed session, a stale/losing delete could silently remove 0 child rows (or race) with no error surfaced — unlike the `deleteChildrenWithVersions`-based services elsewhere, which would explicitly throw a concurrent-child-modification exception instead of silently succeeding on a partial cleanup.

File: core/src/main/java/org/apache/gravitino/storage/relational/service/TagMetaService.java [hunk 2]
```diff
@@ -156,27 +174,47 @@ public <E extends Entity & HasIdentifier> TagEntity updateTag(
...
+        () -> deleteTagWithVersion(identifier, tagPO),
         () ->
```
Comment (jerryshao) [inline core/src/main/java/org/apache/gravitino/storage/relational/service/TagMetaService.java]: **test-coverage**: The four new cascade cleanup calls this PR wires into `deleteTag` — for policy-tag relations, policy-metadata-object relations, owner relations, and securable-object relations — have no test assertions anywhere in the diff.

CLAUDE.md: "Write unit tests for ALL new logic. NO tests = NO merge." The new test `testStaleTagDeleteRollsBackRelationshipCleanup` only asserts `countActiveTagRel` (the one cascade, tag-metadata-object rel, that already existed pre-PR). A wiring mistake in any of the four newly-added cascade calls — wrong mapper method, wrong id/type argument, or a missing rollback under the OCC-conflict path — would not be caught by CI, and combined with the missing row-count check (see the companion comment on this same method), a bug here would fail silently rather than loudly.

File: core/src/main/java/org/apache/gravitino/storage/relational/service/TagMetaService.java [hunk 3]
```diff
@@ -625,6 +678,66 @@ private static void validateAllowedValue(TagPO tagPO, TagValue tagValue)
...
+            SessionUtils.getWithoutCommit(
+                TagMetaMapper.class,
+                mapper -> mapper.selectTagByTagIdForUpdate(observedTagPO.getTagId())),
+        null,
+        current ->
+            Objects.equals(current.getTagName(), observedTagPO.getTagName())
+                && Objects.equals(current.getMetalakeId(), observedTagPO.getMetalakeId()));
+  }
+
+  private List<TagPO> lockTagsForAssignment(List<TagPO> observedTagPOs) {
```
Comment (jerryshao) [inline core/src/main/java/org/apache/gravitino/storage/relational/service/TagMetaService.java:823]: **reuse**: `lockTagsForAssignment` hand-rolls the identical "select-for-update, then verify null/name/metalakeId identity, then throw `NoSuchEntityException`" logic that `OccWriteSupport.lockParentForChildWrite` already provides — and that this very class calls correctly ~30 lines earlier in `lockMetalakeForTagCreate`.

The loop body reimplements exactly what `OccWriteSupport.lockParentForChildWrite` (already used elsewhere in this same PR and by `CatalogMetaService`/`SchemaMetaService`) does per item. A future fix to the shared classification logic (e.g. an exception-message format change, or adding a check) will silently miss this call site since it isn't routed through the helper, reintroducing the drift `OccWriteSupport` (added by #12639) was built specifically to eliminate. The identical tagName/metalakeId identity predicate is also separately duplicated in `tagWriteFailure` (line 705-717) in the same file.

---

**altitude** (same location): The codebase now has three independent implementations of "lock a tag row and verify its identity is unchanged": `tagWriteFailure`'s predicate, this new `lockTagsForAssignment`, and the pre-existing `PolicyTagRelService.lockTag` (untouched by this PR) — despite this PR actively touching tag-locking logic and having the natural opportunity to consolidate. A future change to the identity-check semantics now has to be found and updated in three places across two files; it's easy to update two and miss the third.

---

**efficiency** (same location): This also locks tags one row at a time in a loop (`selectTagByTagIdForUpdate` per tag) instead of a single batched `SELECT ... WHERE tag_id IN (...) FOR UPDATE`, on top of an earlier batch SELECT (`getTagPOsByMetalakeAndNames`) whose result is discarded and immediately re-fetched row-by-row here. For a metadata object tagged with 10 tags, that's 10+ extra DB round trips serialized inside one transaction. A batched, `FOR UPDATE` variant of the existing `listTagPOsByTagIds`-style query would collapse this to one round trip while preserving the tagId-ascending lock order already used to avoid deadlocks.

---

**test-coverage** (same location): This method's concurrent-modification branch (thrown when a tag being associated was renamed or deleted between the initial name lookup and the locking re-read) also has no dedicated test — none of the new tests in `TestTagMetaService.java`'s diff hunk exercise a tag renamed/deleted mid-`associateTagValuesWithMetadataObject`.

File: core/src/main/java/org/apache/gravitino/storage/relational/service/TagMetaService.java [hunk 4]
```diff
@@ -156,27 +174,47 @@ public <E extends Entity & HasIdentifier> TagEntity updateTag(
...
+    TagPO tagPO;
+    try {
+      tagPO = getTagPOByMetalakeAndName(identifier.namespace().level(0), identifier.name());
+    } catch (NoSuchEntityException e) {
+      return false;
+    }
+    return deleteTag(identifier, tagPO);
+  }
+
+  boolean deleteTag(NameIdentifier identifier, TagPO tagPO) {
```
Comment (jerryshao) [inline core/src/main/java/org/apache/gravitino/storage/relational/service/TagMetaService.java]: **conventions**: New package-private overload `boolean deleteTag(NameIdentifier, TagPO)` is spliced between the public `deleteTag(NameIdentifier)` (line 176) and the public `listTagsForMetadataObject` (line 223), breaking the visibility grouping CLAUDE.md requires.

CLAUDE.md's repo-root rules state: "Class Member Ordering: Follow the order: ... 5. Methods (Group by visibility, putting `private` methods at the end)." This new package-private method sits directly between two public methods, so the sequence becomes public -> package-private -> public instead of all public methods being grouped together — the same class of violation found in the companion Model-OCC PR (#12649).

File: core/src/main/java/org/apache/gravitino/storage/relational/service/TagMetaService.java [hunk 5]
```diff
@@ -625,6 +678,66 @@ private static void validateAllowedValue(TagPO tagPO, TagValue tagValue)
         Arrays.toString(allowedValues));
   }
 
+  void lockMetalakeForTagCreate(MetalakePO observedMetalakePO) {
```
Comment (jerryshao) [inline core/src/main/java/org/apache/gravitino/storage/relational/service/TagMetaService.java]: **test-coverage**: `lockMetalakeForTagCreate`'s `NoSuchEntityException` branch (fired when the parent metalake is renamed or removed concurrently with a tag insert) has no dedicated test.

CLAUDE.md: "Write unit tests for ALL new logic. NO tests = NO merge." None of the three new tests added in `TestTagMetaService.java`'s diff hunk (`testTagAlterDeleteAndOverwriteUseMonotonicVersion`, `testTagAlterReportsOptimisticLockConflict`, `testStaleTagDeleteRollsBackRelationshipCleanup`) exercise a concurrent metalake rename/delete racing a tag `insertTag` call, so a regression in this new locking path would not be caught by CI.

File: core/src/main/java/org/apache/gravitino/storage/relational/service/TagMetaService.java [hunk 6]
```diff
@@ -625,6 +678,66 @@ private static void validateAllowedValue(TagPO tagPO, TagValue tagValue)
...
+        () ->
+            SessionUtils.getWithoutCommit(
+                TagMetaMapper.class,
+                mapper ->
+                    mapper.softDeleteTagMetaByIdAndVersion(
+                        observedTagPO.getTagId(), observedTagPO.getCurrentVersion())),
+        () -> tagWriteFailure(identifier, observedTagPO));
+  }
+
+  private RuntimeException tagWriteFailure(NameIdentifier identifier, TagPO observedTagPO) {
```
Comment (jerryshao) [inline core/src/main/java/org/apache/gravitino/storage/relational/service/TagMetaService.java:803]: **test-coverage**: `tagWriteFailure`'s `NoSuchEntityException` branch (tag genuinely gone/renamed, as opposed to a pure version conflict) has no dedicated test — only the `OptimisticLockException` branch is exercised.

CLAUDE.md: "NO tests = NO merge." `tagWriteFailure` returns `NoSuchEntityException` when the re-read current PO is null or its identity (tagName/metalakeId) differs from the observed one, or `OptimisticLockException` otherwise. The new tests (`testTagAlterReportsOptimisticLockConflict`, `testStaleTagDeleteRollsBackRelationshipCleanup`) only exercise the `OptimisticLockException` path; the identity-mismatch/not-found path is untested, so a bug that mis-classifies a deleted/renamed tag as a plain conflict (or vice versa) would go undetected.

File: core/src/main/java/org/apache/gravitino/storage/relational/service/TagMetaService.java [hunk 7]
```diff
@@ -143,7 +161,7 @@ public <E extends Entity & HasIdentifier> TagEntity updateTag(
                       POConverters.updateTagPOWithVersion(tagPO, updatedTagEntity), tagPO));
 
       if (result == null || result == 0) {
-        throw new IOException("Failed to update the entity: " + identifier);
+        throw tagWriteFailure(identifier, tagPO);
```
Comment (jerryshao) [inline core/src/main/java/org/apache/gravitino/storage/relational/service/TagMetaService.java:155]: **correctness / observability**: `updateTag`'s CAS-failure path now throws `tagWriteFailure`'s unchecked exception types (`NoSuchEntityException`/`OptimisticLockException`) instead of the old checked `IOException`. This makes `TagManager.alterTag`'s `catch (IOException ioe)` block (server/core `core/src/main/java/org/apache/gravitino/tag/TagManager.java:202-205`, which logs at ERROR with tag/metalake context before rethrowing) dead code for the concurrent-alter-conflict case, since that file isn't touched by this diff and the new exception type skips straight past it to the generic REST-layer `ExceptionHandlers` (which does correctly map `OptimisticLockException` to an HTTP conflict, so end users aren't broken).

The operator-facing ERROR log with tag/metalake context is silently bypassed in favor of a more generic WARN logged elsewhere — an observability regression for anyone relying on `TagManager`'s logs to diagnose concurrent-write conflicts, and the now-unreachable `catch (IOException ioe)` block in `TagManager.alterTag` is worth a cleanup pass.

**Review by jerryshao — COMMENTED** [review]
Automated code review (Claude Code) — 10 findings below, most severe first. This PR is notably cleaner than the companion Model-OCC PR (#12649): I specifically checked for the same TOCTOU alias-race bug class found there and it's absent here — this PR correctly uses the shared `OccWriteSupport` helper (from #12639) for its single-entity locking paths. The remaining findings are an undocumented cascade-delete safety gap, several missing tests for new failure branches, one real reuse gap (a hand-rolled duplicate of `OccWriteSupport`), a CLAUDE.md convention violation, and an observability regression.

File: core/src/main/java/org/apache/gravitino/storage/relational/service/TagMetaService.java [hunk 8]
```diff
@@ -625,6 +678,66 @@ private static void validateAllowedValue(TagPO tagPO, TagValue tagValue)
...
+    OccWriteSupport.lockParentForChildWrite(
+        observedMetalakePO.getMetalakeName(),
+        Entity.EntityType.METALAKE,
+        () ->
+            SessionUtils.getWithoutCommit(
+                MetalakeMetaMapper.class,
+                mapper ->
+                    mapper.selectMetalakeMetaByIdForShare(observedMetalakePO.getMetalakeId())),
+        null,
+        current -> Objects.equals(current.getMetalakeName(), observedMetalakePO.getMetalakeName()));
```
[the remark this answers is not in the document]
Comment (yuqi1129) [reply core/src/main/java/org/apache/gravitino/storage/relational/service/TagMetaService.java:727]: Handled by the shared helper: `OccWriteSupport.lockParentForChildWrite` returns/throws on `currentParent == null` before it ever calls the predicate, so a metalake that is concurrently deleted produces a `NoSuchEntityException`, not an NPE. That branch is covered by `TestOccWriteSupport.testLockParentForChildWriteNotFoundThrows`.

File: core/src/main/java/org/apache/gravitino/storage/relational/service/TagMetaService.java [hunk 9]
```diff
@@ -625,6 +678,66 @@ private static void validateAllowedValue(TagPO tagPO, TagValue tagValue)
...
+    return OccWriteSupport.writeFailure(
+        identifier,
+        Entity.EntityType.TAG,
+        () ->
+            SessionUtils.getWithoutCommit(
+                TagMetaMapper.class,
+                mapper -> mapper.selectTagByTagIdForUpdate(observedTagPO.getTagId())),
+        null,
+        current ->
+            Objects.equals(current.getTagName(), observedTagPO.getTagName())
```
[the remark this answers is not in the document]
Comment (yuqi1129) [reply core/src/main/java/org/apache/gravitino/storage/relational/service/TagMetaService.java:813]: Same shape as the other one: `OccWriteSupport.writeFailure` checks `currentPO == null` first and only then applies the predicate, so a concurrently deleted tag is classified as `NoSuchEntityException`. Covered by `TestOccWriteSupport.testWriteFailureReturnsNoSuchEntityWhenNotFound`.

File: core/src/main/java/org/apache/gravitino/storage/relational/service/TagMetaService.java [hunk 10]
```diff
@@ -625,6 +678,66 @@ private static void validateAllowedValue(TagPO tagPO, TagValue tagValue)
...
+          || !Objects.equals(lockedTagPO.getMetalakeId(), observedTagPO.getMetalakeId())) {
+        throw new NoSuchEntityException(
+            NoSuchEntityException.NO_SUCH_ENTITY_MESSAGE,
+            Entity.EntityType.TAG.name().toLowerCase(),
+            observedTagPO.getTagName());
+      }
+      lockedTagPOs.add(lockedTagPO);
+    }
+    return lockedTagPOs;
+  }
```
[the remark this answers is not in the document]
Comment (yuqi1129) [reply core/src/main/java/org/apache/gravitino/storage/relational/service/TagMetaService.java:866]: Fixed. The tags are now locked by one statement, `listTagPOsByTagIdsForUpdate`, which selects `WHERE tag_id IN (...) ORDER BY tag_id FOR UPDATE`; the returned rows are then validated against the observed set. That keeps the tag-ID lock order and collapses the round trips to one.

File: core/src/test/java/org/apache/gravitino/storage/relational/service/TestTagMetaService.java [hunk 11]
```diff
@@ -321,6 +327,124 @@ public void testUpdateTag() throws IOException {
     Assertions.assertEquals(tagEntity2, loadedTagEntity1);
   }
 
+  @TestTemplate
+  public void testTagAlterDeleteAndOverwriteUseMonotonicVersion() throws IOException {
+    createAndInsertMakeLake(METALAKE_NAME);
+    TagMetaService tagMetaService = TagMetaService.getInstance();
```
[the remark this answers is not in the document]
Comment (yuqi1129) [reply core/src/test/java/org/apache/gravitino/storage/relational/service/TestTagMetaService.java:350]: The description does list the commands under "How was this patch tested?"; it has been updated with the suites run for this round (`TestTagMetaService`, `TestPolicyTagRelService`, `TestTagManager`).

File: core/src/main/java/org/apache/gravitino/storage/relational/service/TagMetaService.java (hunk 1 above)
Comment (yuqi1129) [reply core/src/main/java/org/apache/gravitino/storage/relational/service/TagMetaService.java]: Documented rather than restructured. `deleteTag` now carries a comment stating that the version-checked delete must stay first, because that is what decides whether the delete wins and a losing delete rolls the transaction back before any dependent row is touched.

`OccWriteSupport.deleteChildrenWithVersions` does not fit these five calls: it asserts that the affected row count equals the size of a known child list, while these are blanket soft deletes for one tag ID where zero rows is a legitimate outcome (a tag with no assignments, no policies, no owner and no securable object). The comment says that too, so the next reader does not reach for the helper by reflex.

File: core/src/main/java/org/apache/gravitino/storage/relational/service/TagMetaService.java (hunk 2 above)
Comment (yuqi1129) [reply core/src/main/java/org/apache/gravitino/storage/relational/service/TagMetaService.java]: Added `testDeleteTagCleansEveryDependentRelation`: it sets up an assignment on a catalog, a policy-tag relation, a policy applied to the tag itself, an owner relation and a role whose securable object names the tag, asserts one active row in each of the five tables, deletes the tag and asserts all five are gone. A wrong mapper method or a wrong id/type argument in any of the four new cascade calls now fails CI.

File: core/src/main/java/org/apache/gravitino/storage/relational/service/TagMetaService.java (hunk 3 above)
Comment (yuqi1129) [reply core/src/main/java/org/apache/gravitino/storage/relational/service/TagMetaService.java:823]: All three points addressed.

reuse/altitude: `lockTagsForAssignment` and `PolicyTagRelService.lockTag` both route through a new `TagMetaService.lockTags`, which classifies through `OccWriteSupport.lockParentForChildWrite`. The three hand-rolled copies are now one call site of the shared helper.

efficiency: `lockTags` issues a single `listTagPOsByTagIdsForUpdate` (`WHERE tag_id IN (...) ORDER BY tag_id FOR UPDATE`) and matches the locked rows against the observed ones, so a metadata object with 10 tags costs one round trip instead of ten, with the same tag-ID lock order.

test-coverage: the missing-row classification is covered by `testDeleteOfAlreadyDeletedTagReportsMissingTagNotConflict` and by `TestOccWriteSupport`. A tag renamed by another transaction in the middle of this one still is not covered: it needs two connections interleaved inside one transaction, which this suite has no harness for.

File: core/src/main/java/org/apache/gravitino/storage/relational/service/TagMetaService.java (hunk 4 above)
Comment (yuqi1129) [reply core/src/main/java/org/apache/gravitino/storage/relational/service/TagMetaService.java]: Fixed. The package-private overload now sits below the public methods, before the private helpers.

File: core/src/main/java/org/apache/gravitino/storage/relational/service/TagMetaService.java (hunk 5 above)
Comment (yuqi1129) [reply core/src/main/java/org/apache/gravitino/storage/relational/service/TagMetaService.java]: Added `testTagCreateIsFencedByParentMetalake`, which asserts both `insertTag(tag, false)` and `insertTag(tag, true)` fail with `NoSuchEntityException` when the parent metalake is not there.

The narrower branch where the metalake is renamed or dropped by another transaction between the lookup and the locking read needs two interleaved connections, which this suite cannot express; the classification itself is covered by `TestOccWriteSupport.testLockParentForChildWriteNotFoundThrows` and `...IdentityMismatchThrows`.

File: core/src/main/java/org/apache/gravitino/storage/relational/service/TagMetaService.java (hunk 6 above)
Comment (yuqi1129) [reply core/src/main/java/org/apache/gravitino/storage/relational/service/TagMetaService.java:803]: Added `testDeleteOfAlreadyDeletedTagReportsMissingTagNotConflict`: it captures the observed row, deletes the tag, then replays the delete with the stale row and asserts `NoSuchEntityException` rather than `OptimisticLockException`, which is exactly the mis-classification you describe.

File: core/src/main/java/org/apache/gravitino/storage/relational/service/TagMetaService.java (hunk 7 above)
Comment (yuqi1129) [reply core/src/main/java/org/apache/gravitino/storage/relational/service/TagMetaService.java:155]: Good catch on the observability gap. `TagManager.alterTag` and `deleteTag` now catch `OptimisticLockException`, log a WARN naming the tag and the metalake, and rethrow, so the operator-facing record survives the change of exception type. The `catch (IOException ioe)` blocks stay: they are still reachable for store failures that are not conflicts.

File: core/src/main/java/org/apache/gravitino/storage/relational/service/TagMetaService.java (hunk 3 above)
Comment (yuqi1129) [reply core/src/main/java/org/apache/gravitino/storage/relational/service/TagMetaService.java:823]: Follow-up after the final concurrency polish: `testTagAssignmentWaitsForConcurrentRenameAndRollsBack` now exercises the previously missing two-connection case. It keeps the rename uncommitted, verifies assignment waits on the batched tag lock, then commits and asserts the assignment fails as not-found with no relation row left behind.

File: core/src/main/java/org/apache/gravitino/storage/relational/service/TagMetaService.java (hunk 5 above)
Comment (yuqi1129) [reply core/src/main/java/org/apache/gravitino/storage/relational/service/TagMetaService.java]: Follow-up after the final concurrency polish: this is now covered by a real two-connection interleaving test, `testTagCreateWaitsForConcurrentParentDelete`. It holds the parent delete uncommitted, verifies create is blocked on the shared parent lock, then commits the delete and asserts `NoSuchEntityException` plus no orphan tag root.

File: core/src/main/java/org/apache/gravitino/storage/relational/service/TagMetaService.java (hunk 6 above)
Comment (yuqi1129) [reply core/src/main/java/org/apache/gravitino/storage/relational/service/TagMetaService.java:803]: Follow-up after the contract audit: I corrected this path to preserve the existing idempotent delete contract. If the stable ID is gone or its natural key changed, the delete returns false; if the same tag still exists with a newer version, it throws OptimisticLockException. The stale-same-identity and already-deleted cases are covered separately.

**Review by jerryshao — APPROVED** [review]
